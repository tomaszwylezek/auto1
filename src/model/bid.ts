export interface IBid {
  id: string,
  carTitle: string,
  amount: number,
  created: string
}
