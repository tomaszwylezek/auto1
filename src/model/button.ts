export enum ButtonTheme {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  danger = 'danger',
  warning = 'warning',
  info = 'info',
  light = 'light',
  dark = 'dark',
  default = 'default',
  outlinePrimary = 'outline-primary',
  outlineSecondary = 'outline-secondary',
  outlineSuccess = 'outline-success',
  outlineDanger = 'outline-danger',
  outlineWarning = 'outline-warning',
  outlineInfo = 'outline-info',
  outlineLight = 'outline-light',
  outlineDark = 'outline-dark'
}

export enum Size {
  sm = 'sm',
  lg = 'lg'
}
