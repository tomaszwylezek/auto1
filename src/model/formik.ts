import React from 'react';

export interface FormikField {
  onChange: (e: React.ChangeEvent<any>) => void;
  onBlur: (e: any) => void;
  value: any;
  name: string;
}
