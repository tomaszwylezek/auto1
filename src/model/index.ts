export * from './store';
export * from './merchant';
export * from './bid';
export * from './formik';
