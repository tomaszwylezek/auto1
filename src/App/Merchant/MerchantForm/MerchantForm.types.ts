import { IMerchant } from 'model';

export interface IMerchantFormOwnProps {
  onSubmit: (values: any) => void;
  merchant?: IMerchant;
}

export interface IMerchantFormProps extends IMerchantFormOwnProps {}

export type IMerchantFormValues = Omit<IMerchant, 'bids' | 'id'>;
