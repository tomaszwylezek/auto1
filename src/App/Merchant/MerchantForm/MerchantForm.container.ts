import * as Yup from 'yup';
import { compose } from 'redux';
import { FormikBag, withFormik } from 'formik';
import { MerchantForm } from './MerchantForm';

import { IMerchant } from 'model';
import { IMerchantFormProps, IMerchantFormValues } from 'App/Merchant/MerchantForm/MerchantForm.types';

const initialFormValues: IMerchantFormValues = {
  firstname: '',
  lastname: '',
  phone: '',
  avatarUrl: '',
  email: '',
  hasPremium: false
};

const getMerchantValues: (area: IMerchant) => IMerchantFormValues = ({
  firstname,
  lastname,
  phone,
  avatarUrl,
  email,
  hasPremium
}: IMerchant): IMerchantFormValues => ({
  firstname,
  lastname,
  phone,
  avatarUrl,
  email,
  hasPremium
});

export const MerchantFormContainer = compose(
  withFormik<IMerchantFormProps, IMerchantFormValues>({
    displayName: 'MerchantFormContainer',
    handleSubmit: (
      values: IMerchantFormValues,
      { props: { onSubmit } }: FormikBag<IMerchantFormProps, IMerchantFormValues>
    ): void => onSubmit(values),
    mapPropsToValues: ({ merchant }) => (merchant ? getMerchantValues(merchant) : initialFormValues),
    validationSchema: () =>
      Yup.object().shape<Partial<IMerchantFormValues>>({
        firstname: Yup.string().required(),
        lastname: Yup.string().required(),
        email: Yup.string().email()
      })
  })
)(MerchantForm);
