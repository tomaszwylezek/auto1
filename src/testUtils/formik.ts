import { FormikActions, FormikHandlers, FormikProps } from 'formik';
import { FormikField } from 'model';

export function getMockedFormikConfig<T>(options: Partial<FormikProps<T>> = {}): FormikProps<T> {
  return {
    ...getMockedFormikActions(),
    ...getMockedFormikHandles(),
    enableReinitialize: false,
    isValidating: false,
    isSubmitting: false,
    dirty: false,
    isValid: false,
    submitCount: 0,
    registerField: jest.fn(),
    unregisterField: jest.fn(),
    values: {} as T,
    errors: {} as T,
    initialValues: {} as T,
    touched: {} as T,
    ...options
  };
}

export function getMockedFormikActions<V>(options: Partial<FormikActions<V>> = {}): FormikActions<V> {
  return {
    setStatus: jest.fn(),
    setError: jest.fn(),
    setErrors: jest.fn(),
    setSubmitting: jest.fn(),
    setTouched: jest.fn(),
    setValues: jest.fn(),
    setFieldValue: jest.fn(),
    setFieldError: jest.fn(),
    setFieldTouched: jest.fn(),
    validateForm: jest.fn(),
    validateField: jest.fn(),
    resetForm: jest.fn(),
    submitForm: jest.fn(),
    setFormikState: jest.fn(),
    ...options
  };
}

export function getMockedFormikHandles(options: Partial<FormikHandlers> = {}): FormikHandlers {
  return {
    handleSubmit: jest.fn(),
    handleReset: jest.fn(),
    handleBlur: jest.fn(),
    handleChange: jest.fn(),
    ...options
  };
}

export function getMockedField(options: Partial<FormikField> = {}): FormikField {
  return {
    onChange: jest.fn(),
    onBlur: jest.fn(),
    value: '',
    name: '',
    ...options
  };
}
