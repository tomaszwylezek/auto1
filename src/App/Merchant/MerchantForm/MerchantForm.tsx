import React, { FC } from 'react';
import { IMerchantFormProps, IMerchantFormValues } from 'App/Merchant/MerchantForm/MerchantForm.types';
import { Field, Form, FormikProps } from 'formik';
import { Input } from 'shared/form/Input/Input';
import { Checkbox } from 'shared/form/Checkbox/Checkbox';
import { Button } from 'shared/Button/Button';

// Models

export const MerchantForm: FC<IMerchantFormProps & FormikProps<IMerchantFormValues>> = ({ merchant }) => {
  const isEdit: boolean = !!merchant;
  return (
    <Form className="row">
      <div className="col-12">
        <Field component={Input} name="firstname" label="First name" />
      </div>
      <div className="col-12">
        <Field component={Input} name="lastname" label="Last name" />
      </div>
      <div className="col-12">
        <Field component={Input} name="avatarUrl" label="Avatar Url" />
      </div>
      <div className="col-12">
        <Field component={Input} name="email" label="Email" />
      </div>
      <div className="col-12">
        <Field component={Input} name="phone" label="Phone" />
      </div>
      <div className="col-12">
        <Field component={Checkbox} name="hasPremium" label="Has premium" />
      </div>
      <div className="col-12">
        <Button type="submit">{isEdit ? 'Edit' : 'Add'}</Button>
      </div>
    </Form>
  );
};
