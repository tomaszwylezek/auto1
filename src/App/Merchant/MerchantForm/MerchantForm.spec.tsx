import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { MerchantForm } from './MerchantForm';
import { getMockedFormikConfig } from 'testUtils/formik';

// Models
import { IMerchantFormProps, IMerchantFormValues } from './MerchantForm.types';
import { FormikProps } from 'formik';
import { IMerchant } from 'model';

describe('MerchantForm component should', (): void => {
  let props: IMerchantFormProps & FormikProps<IMerchantFormValues>;
  let component: ShallowWrapper<IMerchantFormProps & FormikProps<IMerchantFormValues>>;

  beforeEach((): void => {
    props = {
      onSubmit: jest.fn(),
      ...getMockedFormikConfig()
    };

    component = shallow(<MerchantForm {...props} />);
  });

  it('render correctly when no merchant', (): void => {
    expect(component).toMatchSnapshot();
  });

  it('change button text when merchant available', (): void => {
    component.setProps({
      merchant: {} as IMerchant
    });

    expect(component.find('Button')).toMatchSnapshot('Edit');
  });

  it('change button to not disable when valid form', (): void => {
    component.setProps({
      isValid: true,
      dirty: true
    });

    expect(component.find('Button')).toMatchSnapshot('Edit');
  });
});
