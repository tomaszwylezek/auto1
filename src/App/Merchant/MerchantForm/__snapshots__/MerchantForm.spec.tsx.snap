// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerchantForm component should change button text when merchant available: Edit 1`] = `
<Button
  disabled={true}
  type="submit"
>
  Edit
</Button>
`;

exports[`MerchantForm component should change button to not disable when valid form: Edit 1`] = `
<Button
  disabled={false}
  type="submit"
>
  Add
</Button>
`;

exports[`MerchantForm component should render correctly when no merchant 1`] = `
<Form
  className="row"
>
  <div
    className="col-12"
  >
    <FormikConnect(FieldInner)
      component={[Function]}
      label="First name"
      name="firstname"
      required={true}
    />
  </div>
  <div
    className="col-12"
  >
    <FormikConnect(FieldInner)
      component={[Function]}
      label="Last name"
      name="lastname"
      required={true}
    />
  </div>
  <div
    className="col-12"
  >
    <FormikConnect(FieldInner)
      component={[Function]}
      label="Email"
      name="email"
      required={true}
    />
  </div>
  <div
    className="col-12"
  >
    <FormikConnect(FieldInner)
      component={[Function]}
      label="Avatar Url"
      name="avatarUrl"
    />
  </div>
  <div
    className="col-12"
  >
    <FormikConnect(FieldInner)
      component={[Function]}
      label="Phone"
      name="phone"
    />
  </div>
  <div
    className="col-12"
  >
    <FormikConnect(FieldInner)
      component={[Function]}
      label="Has premium"
      name="hasPremium"
    />
  </div>
  <div
    className="col-12"
  >
    <Button
      disabled={true}
      type="submit"
    >
      Add
    </Button>
  </div>
</Form>
`;
